chore: test loading multi accounts

tommasini opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
# 🎯 Performance Quality Gates

This document outlines the performance thresholds (quality gates) for MetaMask Mobile's critical user flows. These thresholds ensure optimal user experience across different platforms and scenarios.

## 📱 Platform Overview

- **Android**: Generally has higher thresholds due to platform constraints
- **iOS**: Lower thresholds leveraging platform optimizations
- **Quality Gates**: Hard limits that cause test failures if exceeded

## 📊 Test Reporting

Performance tests automatically generate detailed reports using the `PerformanceTestReporter` utility:
- **JSON Reports**: Structured performance data for analysis
- **Test Results**: Include timing metrics, thresholds, and pass/fail status
- **User Profile Testing**: Tests run across different user states (CORE_USER, POWER_USER)

---

## 🏠 Account List Performance Tests

### Test: `render account list efficiently with multiple accounts and networks`
**Configuration**: Multiple accounts, popular networks, profile syncing enabled

| Platform | Total Max Time | Notes |
|----------|----------------|--------|
| **Android** | 5000ms | 5 seconds maximum |
| **iOS** | 7,500ms | 7.5 seconds maximum |

---

### Test: `handle account list performance with heavy token load`
**Configuration**: Multiple accounts, popular networks, 10 tokens for stress testing

| Platform | Total Max Time | Notes |
|----------|----------------|--------|
| **Android** | 5000ms | 5 seconds maximum |
| **iOS** | 7,500ms | 7.5 seconds maximum |

---

### Test: `benchmark account list with minimal load`
**Configuration**: Minimal accounts, default network, 2 tokens (baseline measurement)

| Platform | Total Max Time | Notes |
|----------|----------------|--------|
| **Android** | 45,000ms | 45 seconds maximum |
| **iOS** | 15,000ms | 15 seconds maximum |

---

### Test: `benchmark switching accounts from the account list`
**Configuration**: Account switching/dismissal performance

| Platform | Dismissal Max Time | Notes |
|----------|-------------------|--------|
| **Android** | 5,000ms | 5 seconds maximum |
| **iOS** | 4,000ms | 4 seconds maximum |

---

## 🌐 Network List Performance Tests

### Test: `render network list efficiently with multiple accounts and all popular networks`
**Configuration**: Multiple accounts, all popular networks

| Platform | Total Max Time | Notes |
|----------|----------------|--------|
| **Android** | 17,500ms | 17.5 seconds maximum |
| **iOS** | 6,500ms | 6.5 seconds maximum |

---

### Test: `handle network list performance with heavy token load on all popular networks`
**Configuration**: Multiple accounts, popular networks, 10 tokens for stress testing

| Platform | Total Max Time | Notes |
|----------|----------------|--------|
| **Android** | 17,500ms | 17.5 seconds maximum |
| **iOS** | 6,500ms | 6.5 seconds maximum |

---

### Test: `benchmark network list with minimal load`
**Configuration**: Minimal tokens, popular networks (baseline measurement)

| Platform | Render Max Time | Notes |
|----------|----------------|--------|
| **Android** | 2,500ms | 2.5 seconds maximum |
| **iOS** | 1,500ms | 1.5 seconds maximum |

---

### Test: `benchmark switching networks from the network list`
**Configuration**: Network switching/dismissal performance

| Platform | Dismissal Max Time | Notes |
|----------|-------------------|--------|
| **Android** | 2,500ms | 2.5 seconds maximum |
| **iOS** | 1,500ms | 1.5 seconds maximum |

---

## 📊 Test Summary

### Account List Tests (4 tests)
- ✅ **Standard Load**: Multiple accounts, popular networks
- ✅ **Heavy Load**: Multiple accounts, 10 tokens, popular networks  
- ✅ **Baseline Test**: Minimal accounts, 2 tokens, default network
- ✅ **Dismissal Test**: Account switching performance

### Network List Tests (4 tests)
- ✅ **Standard Load**: Multiple accounts, popular networks
- ✅ **Heavy Load**: Multiple accounts, 10 tokens, popular networks
- ✅ **Baseline Test**: Minimal tokens, popular networks
- ✅ **Dismissal Test**: Network switching performance

**Total**: 8 performance tests across critical user flows

---

## 🚨 Quality Gate Rules

### Failure Criteria
Tests fail immediately when total time exceeds the maximum acceptable time for any scenario.

### Performance Patterns
- **Heavy Token Load**: Increases render times but maintains same thresholds as standard load
- **Platform Differences**: iOS consistently performs better than Android
- **Baseline Tests**: Should complete quickly but have generous thresholds for stability
- **Dismissal Tests**: Focus on UI responsiveness during state transitions

### User Profile Testing
Tests run across different user states with varying account complexity:

#### User Profile Definitions
- **CASUAL_USER**: 2 EVM accounts from 1 SRP *(currently not used in tests)*
- **CORE_USER**: 5 EVM accounts and 5 Solana accounts from 1 SRP
- **POWER_USER**: 15 EVM accounts from 2 SRPs + 5 Solana accounts

#### Current Test Coverage
- **CORE_USER**: Standard user configuration for baseline performance
- **POWER_USER**: Enhanced user configuration with maximum account complexity

---

## 📈 Reporting Features

### Automated Reports
- **JSON Output**: Structured performance data for CI/CD integration
- **Test Metrics**: Total time measurements and performance thresholds
- **Threshold Tracking**: Pass/fail status with actual vs. expected performance
- **User Profile Results**: Separate results for different user configurations

### Report Usage
Performance reports can be used for:
- Continuous integration quality gates
- Performance regression detection
- Platform-specific optimization insights
- User experience benchmarking

---


**What can we do to improve it**
* Instead of measuring the action once, we perform the action multiple times and we get an average and the top duration, both need to have base lines.
* To get more granular function reports, we can use this sample work done to measure multiple functions of one flow, for example if add account if it's split in 2 functions, instead of one big measure, we can measure both functions, and get more deep knowledge of where we can improve  https://github.com/MetaMask/metamask-mobile/pull/16806/


**Bitrise Runs**
* https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/05c14c4e-4868-43b0-9642-0e5d69ebd82b?tab=workflows
* https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/9b63202e-8944-46b5-8867-ab94f748f37f
* https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/98ca0cf1-edfa-4888-9e38-ef64422d6833
* https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/7857fee5-3f27-41a6-90a9-0332e1a81435


Table if 8 bitrise runs results
| **Test Name**                                           | **Platform** | **Total Time (Runs)**                                                                 | **Min** | **Max** | **Delta** | **Average** |
|---------------------------------------------------------|--------------|--------------------------------------------------------------------------------------|---------|---------|-----------|-------------|
| render account list efficiently with multiple accounts and networks | IOS          | 2.54, 2.61, 3.65, 2.46, 2.88, 3.1, 2.94, 2.87                                       | 2.46    | 3.65    | 1.19      | 2.88        |
| render account list efficiently with multiple accounts and networks | ANDROID      | 2.83, 2.64, 2.92, 2.28, 2.62, 2.76, 1.95, 2.37                                      | 1.95    | 2.92    | 0.97      | 2.55        |
| handle account list performance with heavy token load   | IOS          | 4.01, 5.72, 4.02, 5.3, 2.49, 2.48, 2.67, 2.78                                       | 2.48    | 5.72    | 3.24      | 3.93        |
| handle account list performance with heavy token load   | ANDROID      | 2.15, 2.51, 1.97, 2.78, 2.13, 2.29, 2.23, 2.41                                      | 1.97    | 2.78    | 0.81      | 2.29        |
| benchmark account list with minimal load               | IOS          | 2.86, 3.48, 2.87, 3.32, 2.81, 3.35, 2.85, 3.13                                       | 2.81    | 3.48    | 0.67      | 3.08        |
| benchmark account list with minimal load               | ANDROID      | 2.08, 2.17, 2.59, 2.28, 2.07, 2.19, 2.13, 2.49                                      | 2.07    | 2.59    | 0.52      | 2.25        |
| benchmark switching networks from the network list      | IOS          | 6.61, 8.11, 7.65, 8.08, 7.64, 6.53, 6.49, 6.55                                       | 6.49    | 8.11    | 1.62      | 7.34        |
| benchmark switching networks from the network list      | ANDROID      | 7.68, 4.26, 4.11, 5.6, 4.04, 4.32, 4.11, 4.28                                       | 4.04    | 7.68    | 3.64      | 4.92        |
| benchmark switching accounts from the account list      | IOS          | 5.99, 4.37, 3.14, 4.55, 3.36, 4.33                                                  | 3.14    | 5.99    | 2.85      | 4.29        |
| benchmark switching accounts from the account list      | ANDROID      | 2.01, 2.28, 1.71, 2.05, 1.77, 2.28, 1.75, 2.39                                      | 1.71    | 2.39    | 0.68      | 2.06        |

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

tommasini committed:
account list load testing

cortisiko committed:
try adding multiple accounts

cortisiko committed:
add withMutipleAccounts fixture builder. solana account is proving to be problematic.

tommasini committed:
baselines for account list rendering

tommasini committed:
add performance to smoke e2e pipeline

tommasini committed:
merge main and solve conflicts

cortisiko committed:
add changes to fix popular network filter. Also extend test to switch between long network list

tommasini committed:
baselines for both OS

cortisiko committed:
run only 1 test on ci for performance workflow

cortisiko committed:
add networklist measurements. + fix account list test on android

cortisiko committed:
fixture for adding tokens to all popular networks

cortisiko committed:
change accounts

tommasini committed:
accounts with tokens with popular networks fixture

tommasini committed:
checksum token addresses and populate erc20 balances in all popular networks

cortisiko committed:
clean up fixture name and adding tokens to wallet with check summned addresses

tommasini committed:
test snap controller state

tommasini committed:
quality gates markdown file

FrederikBolding commented on e2e/fixtures/constants.ts:
```diff
@@ -0,0 +1,462 @@
+      removable: false,
+      sourceCode:
+        '(()=>{var e={251:(e,t)=>{t.read=function(e,t,r,n,i){var s,o,a=8*i-n-1,u=(1<<a)-1,c=u>>1,d=-7,f=r?i-1:0,l=r?-1:1,h=e[t+f];for(f+=l,s=h&(1<<-d)-1,h>>=-d,d+=a;d>0;s=256*s+e[t+f],f+=l,d-=8);for(o=s&(1<<-d)-1,s>>=-d,d+=n;d>0;o=256*o+e[t+f],f+=l,d-=8);if(0===s)s=1-c;else{if(s===u)return o?NaN:1/0*(h?-1:1);o+=Math.pow(2,n),s-=c}return(h?-1:1)*o*Math.pow(2,s-n)},t.write=function(e,t,r,n,i,s){var ...
+      status: 'running',
```
```suggestion
      status: 'stopped',
```

FrederikBolding commented on e2e/fixtures/constants.ts:
```diff
@@ -0,0 +1,462 @@
+      preinstalled: true,
+      removable: false,
+      sourceCode:
+        '(()=>{var e={33:(e,t,n)=>{var r=n(9968),o=n(9756),i=[["ary",128],["bind",1],["bindKey",2],["curry",8],["curryRight",16],["flip",512],["partial",32],["partialRight",64],["rearg",256]];e.exports=function(e,t){return r(i,(function(n){var r="_."+n[0];t&n[1]&&!o(e,r)&&e.push(r)})),e.sort()}},88:(e,t,n)=>{var r=n(4700);e.exports=function(e){return r(this,e).get(e)}},125:(e,t,n)=>{var r=n(9032) ...
```
```suggestion
      status: 'stopped',
```
When loading the app these aren't running right away

FrederikBolding commented on e2e/fixtures/fixture-builder.js:
```diff
@@ -1105,42 +1127,134 @@ class FixtureBuilder {
+          type: 'HD Key Tree',
+          accounts: [
+            '0x43e1c289177ecfbe6ef34b5fb2b66ebce5a8e05b',
+            // Unknown why this 3 addresses are not loading, we should fix it
```
Check that the vault has the correct number of accounts

tommasini committed:
start creating account profiles for performance testing

cortisiko committed:
create user profiles

tommasini committed:
snap status must be stopped so balance can be fetched

tommasini committed:
performance e2e pipeline called smoke_e2e_performance_pipeline

cortisiko committed:
make the fixture methods more generic

cortisiko committed:
clean up fixture builder and remove unused methods

cortisiko committed:
fix tests so they run based on user profile.

cortisiko committed:
generate report for performance tests since bitrise logs are redacted.

cortisiko committed:
use dynamic port

tommasini committed:
update test with rates and balances for erc20 tokens

tommasini committed:
fix: total time (#17049)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

cortisiko committed:
clean up tests. it should be passing on CI now.

cortisiko committed:
create new workflows :( for the performance tests because they will be using artifacts outside of the regular e2e workflows.

cortisiko committed:
fix performance artifact path

cortisiko committed:
use the proper fixture method to run tests

cortisiko committed:
only use 5 tokens instead of 50

cortisiko committed:
use 10 tokens as a test

cortisiko committed:
use balances on all user profiles not just the core user profile

tommasini committed:
remove render and navigation time and format to everything have a totalTime instead, write everything in miliseconds and do not break the tests only create warnings for now

tommasini committed:
update quality-gates markdown file

tommasini committed:
add information about user profile definition in the quality gates markdown file

tommasini committed:
update baselines

tommasini committed:
merge main and solve conflicts

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
merge main and solve conflicts

tommasini commented on e2e/specs/performance/network-list/dismiss-network-list.spec.ts:
```diff
@@ -0,0 +1,124 @@
+
+            const startTime = Date.now();
+            await NetworkListModal.changeNetworkTo('Polygon Mainnet', false);
+            await Assertions.checkIfNotVisible(NetworkListModal.networkScroll);
```
Yes, we can add it! That sounds good!

tommasini committed:
merge main and solve conflicts

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
switch throw new error by console.warn

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
add check for cache stage to e2e performance pipeline

tommasini committed:
merge main and solve confclits

tommasini committed:
nit pick around quality gates commennt and documentation

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
prettier on jest.config file

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

Cal-L reviewed (approved):
LGTM!

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
merge main and solve conflicts

Cal-L reviewed (dismissed):
LGTM!

Cal-L reviewed (dismissed):
LGTM

tommasini committed:
test

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

Cal-L reviewed (dismissed):
Lgtm

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
fix anvil manager

tommasini committed:
fix default acocunt fixture builder

tommasini committed:
fix remaining e2e

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

Cal-L reviewed (dismissed):
LGTM

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
merge main and solve conficts

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
fix anvil manager

tommasini committed:
revert bitrise yml change

tommasini committed:
add steps to copy the performance reports and add revert last commit

tommasini committed:
only run those performance copy results and deploy results on smoke performance

tommasini committed:
replace countains function for eq

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

tommasini committed:
Merge branch 'main' into test-loading-multi-accounts

MarioAslau reviewed (approved):
LGTM
